### PR TITLE
Refs[TASK][37982] Lấy dữ liệu danh sách tướng

### PR DESCRIPTION
**Pull request by hhd0307**
## Related Tickets
- [#Task 37982: Lấy dữ liệu danh sách tướng](https://edu-redmine.sun-asterisk.vn/issues/37982)
## WHAT
- Lấy dữ liệu danh sách tướng
## Review Checklist

Category | View Point | Description | Self review | Reviewer2 (name)
--- | --- | --- | --- | ---
Conventions | Sun* coding conventions | https://github.com/framgia/coding-standards/tree/master/eng/android |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Conventions | Kotlin coding conventions | https://kotlinlang.org/docs/coding-conventions.html |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Redmine | Sun* Redmine working process  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md |<li>- [ ] yes</li>|<li>- [ ] yes</li>  

## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/sun/hero_02/screen/champions/ChampionsFragment.kt [hunk 1]
```diff
@@ -1,8 +1,12 @@
 package com.sun.hero_02.screen.champions
 
 import android.view.LayoutInflater
+import android.widget.Toast
```
Comment (khoaqh-2470) [inline app/src/main/java/com/sun/hero_02/screen/champions/ChampionsFragment.kt]:
```suggestion
```

File: app/src/main/java/com/sun/hero_02/screen/champions/ChampionsFragment.kt [hunk 2]
```diff
@@ -1,8 +1,12 @@
...
+import androidx.lifecycle.Observer
+import com.sun.hero_02.R
 import com.sun.hero_02.base.BaseFragment
 import com.sun.hero_02.databinding.FragmentChampionsBinding
+import com.sun.hero_02.utils.Status
```
Comment (khoaqh-2470) [inline app/src/main/java/com/sun/hero_02/screen/champions/ChampionsFragment.kt]:
```suggestion
```

File: app/src/main/java/com/sun/hero_02/screen/champions/ChampionsViewModel.kt [hunk 3]
```diff
@@ -1,6 +1,44 @@
 package com.sun.hero_02.screen.champions
 
+import android.util.Log
```
Comment (khoaqh-2470) [inline app/src/main/java/com/sun/hero_02/screen/champions/ChampionsViewModel.kt]:
```suggestion
```

File: app/src/main/java/com/sun/hero_02/data/model/ChampionsResponse.kt [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package com.sun.hero_02.data.model
+
+import com.google.gson.annotations.SerializedName
+
+class ChampionsResponse {
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/hero_02/data/model/ChampionsResponse.kt]: data class

File: app/src/main/java/com/sun/hero_02/data/model/ChampionsResponse.kt [hunk 5]
```diff
@@ -0,0 +1,8 @@
...
+    @SerializedName("data")
+    lateinit var championItems: Map<String, Hero>
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/hero_02/data/model/ChampionsResponse.kt]: không được dùng lateinit ở đây

File: app/src/main/java/com/sun/hero_02/data/repository/HeroRepositoryImpl.kt [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+import com.sun.hero_02.data.model.Hero
+import com.sun.hero_02.data.repository.source.HeroDataSource
+import com.sun.hero_02.utils.DataResult
+
+class HeroRepositoryImpl(
+    private val local: HeroDataSource.Local,
+    private val remote: HeroDataSource.Remote
+) : HeroRepository, BaseRepository() {
+
+    override suspend fun getChampions(): DataResult<MutableList<Hero>> = withResultContext {
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/hero_02/data/repository/HeroRepositoryImpl.kt]:
```suggestion
    override suspend fun getChampions() = withResultContext {
```

File: app/src/main/java/com/sun/hero_02/data/repository/source/remote/HeroRemoteImpl.kt [hunk 7]
```diff
@@ -0,0 +1,12 @@
+package com.sun.hero_02.data.repository.source.remote
+
+import com.sun.hero_02.data.model.ChampionsResponse
+import com.sun.hero_02.data.repository.source.HeroDataSource
+import com.sun.hero_02.data.repository.source.remote.api.ApiService
+
+class HeroRemoteImpl(private val apiService: ApiService) : HeroDataSource.Remote {
+
+    override suspend fun getChampions(): ChampionsResponse {
+        return apiService.getChampions()
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/hero_02/data/repository/source/remote/HeroRemoteImpl.kt]:
```suggestion
    override suspend fun getChampions() = apiService.getChampions()
```

File: app/src/main/java/com/sun/hero_02/data/repository/source/remote/api/ApiService.kt [hunk 8]
```diff
@@ -0,0 +1,10 @@
+package com.sun.hero_02.data.repository.source.remote.api
+
+import com.sun.hero_02.data.model.ChampionsResponse
+import com.sun.hero_02.utils.Constant
+import retrofit2.http.GET
+
+interface ApiService {
+    @GET("${Constant.BASE_VERSION}/${Constant.BASE_LANGUAGE}/${Constant.PATH_LIST_CHAMPION}")
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/hero_02/data/repository/source/remote/api/ApiService.kt]: viết string trực tiếp vào đây luôn không cần khai báo trong constant e.
Và những param dùng chung thì add vào header request default luôn

File: app/src/main/java/com/sun/hero_02/data/repository/source/remote/api/ServiceGenerator.kt [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+        val builder = Retrofit.Builder()
+            .baseUrl(baseUrl)
+            .addConverterFactory(GsonConverterFactory.create())
+            .client(buildOkHttpClient())
+            .build()
+        return builder.create(serviceClass)
+    }
+
+    private fun buildOkHttpClient(): OkHttpClient =
+        OkHttpClient.Builder().apply {
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/hero_02/data/repository/source/remote/api/ServiceGenerator.kt]:
```suggestion
    private fun buildOkHttpClient() = OkHttpClient.Builder().apply {
```

File: app/src/main/java/com/sun/hero_02/di/DataSourceModule.kt [hunk 10]
```diff
@@ -0,0 +1,17 @@
+package com.sun.hero_02.di
+
+import com.sun.hero_02.data.repository.source.HeroDataSource
+import com.sun.hero_02.data.repository.source.local.HeroLocalImpl
+import com.sun.hero_02.data.repository.source.remote.HeroRemoteImpl
+import org.koin.dsl.module
+
+val dataSourceModule = module {
+
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/hero_02/di/DataSourceModule.kt]: remove blank

File: app/src/main/java/com/sun/hero_02/screen/champions/ChampionsViewModel.kt [hunk 11]
```diff
@@ -1,6 +1,44 @@
...
-class ChampionsViewModel : ViewModel() {
+class ChampionsViewModel(private val heroRepository: HeroRepository) : ViewModel() {
+
+    private val _champions = MutableLiveData<MutableList<Hero>>()
+    private val _status = MutableLiveData<Status>()
+
+    val champions: LiveData<MutableList<Hero>>
+        get() = _champions
+    val status: LiveData<Status>
+        get() = _status
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/hero_02/screen/champions/ChampionsViewModel.kt]:
```suggestion
    private val _champions = MutableLiveData<MutableList<Hero>>()
    val champions: LiveData<MutableList<Hero>>
        get() = _champions
        
    private val _status = MutableLiveData<Status>()
    val status: LiveData<Status>
        get() = _status
```

**Review by khoaqh-2470 — APPROVED** [review]

**Review by tuanlvt-1117 — APPROVED** [review]

**Pull request merged**